providers/azure.go

```diff
@@ -233,13 +234,50 @@ func getEmailFromJSON(json *simplejson.Json) (string, error) {
 	return email, err
 }
 
-// GetEmailAddress returns the Account email address
-func (p *AzureProvider) GetEmailAddress(ctx context.Context, s *sessions.SessionState) (string, error) {
+// full claims that can be emitted by AAD are
+// https://docs.microsoft.com/en-us/azure/active-directory/develop/id-tokens
+// only include email here since we only need email extraction
+type aadIDTokenClaims struct {
+	Email string `json:"email,omitempty"`
+}
+
+// Valid is the function validating the claims,
+// required by jwt.Claims interface.
+// returning nil as we only use it to extract claims instead of validating them
+func (aadIDTokenClaims) Valid() error {
+	return nil
+}
+
+func getEmailFromIDToken(idToken string) (string, error) {
+	claims := &aadIDTokenClaims{}
+	parser := jwt.Parser{}
+
+	// ParseUnverified to extract claims without verifying signature
+	if _, _, err := parser.ParseUnverified(idToken, claims); err != nil {
+		return "", err
+	}
+	if claims.Email == "" {
+		return "", errors.New("missing email claim from id_token")
+	}
+	return claims.Email, nil
+}
+
+// EnrichSessionState finds the email to enrich the session state
+func (p *AzureProvider) EnrichSessionState(ctx context.Context, s *sessions.SessionState) error {
 	var email string
-	var err error
+
+	if s.IDToken != "" {
+		email, err := getEmailFromIDToken(s.IDToken)
+		if err != nil || email == "" {
```

Comment:
if `email == ""` is a thing and we don't like it, I feel like it should be handled in `getEmailFromIDToken` and raise an appropriate error there.

Then this can be simplified and clearer to something like this:
```
	if s.IDToken != "" {
		email, err := getEmailFromIDToken(s.IDToken)
		if err == nil {
			s.Email = email
			return nil
		}
		logger.Errorf("unable to find email from id_token: %s", err)
	}
```

Reply:
good suggestion. thanks!